### ISSUE TITLE
Fix S1125: FP with nullable and VS2015/MSBuild 14

### ISSUE THREAD

**Issue by DuAell** [SonarSource/sonar-csharp#1325]
### Description

False positive on rule "Boolean literals should not be redundant" due to nullable bool

Rule key: S1125

### Repro steps

Please provide the steps required to reproduce the problem

```
public bool? IsStateValid { get; set; }

Sockets.All(x => x.IsStateValid == true); // False-positive here
```

### Expected behavior

Rule shouldn't detect an error here

### Related information

* SonarQube version: Version 7.0 (build 36138)

**Comment by Evangelink** [comment]
Hi @DuAell,

Could you provide a more complete example? I cannot have any issue with the following code:

```
class Program
{
    static void Main(string[] args)
    {
        var result = Sockets.All(x => x.IsStateValid == true);
    }

    public static IEnumerable<Socket> Sockets { get; }
}

public class Socket
{
    public bool? IsStateValid { get; set; }
}
```

**Comment by DuAell** [comment]
Can't really do.

IsStateValid is used like that in my code : 

public bool IsValid => Configuration == null || Sockets.All(x => x.IsStateValid == true);

I have to use == true because IsStateValid is nullable, but SonarQube tells me to remove "== true" because of this rule.

**Comment by Evangelink** [comment]
Ok! Can you confirm that the code snippet a posted in my previous answer also doesn't trigger any issue?

**Comment by DuAell** [comment]
Confirmed.

Try that : 

    class Program
    {
        static void Main(string[] args)
        {
            var result = Sockets.All(x => x.IsStateValid == true); // SonarQube finds no issue
        }

        public static IEnumerable<Socket> Sockets { get; }

        public bool IsValid => Sockets.All(x => x.IsStateValid == true); // SonarQube raises false positive issue.
    }

    public class Socket
    {
        public bool? IsStateValid { get; set; }
    }

**Comment by Evangelink** [comment]
Thx for the confirmation. 

I still don't have any issue raised with your code snippet. Which version of SonarC# are you using? Is it through NuGet package or SonarLint?

**Comment by DuAell** [comment]
It's SonarC# running on SonarQube.

SonarC# version : 7.0.1 (build 4822)
SonarQube version : Version 7.0 (build 36138)

**Comment by Evangelink** [comment]
Ok. Still no issue with 7.0.1. I do manage to have an issue when removing the nullable from `IsStateValid`.

I'll try to see from the code if I can spot which usage could lead to the issue. In the meantime if you manage to get a piece of code reproducing the error feel free to share.

**Comment by Evangelink** [comment]
That's indeed pretty weird because neither me nor my colleagues are able to reproduce the issue with `SonarC# 7.0.1` and the code snippet you provided.

**Comment by DuAell** [comment]
![image](https://user-images.githubusercontent.com/708163/38817408-974ae2b8-4198-11e8-9a09-b59f1b54be7a.png)

**Comment by Evangelink** [comment]
I finally managed to reproduce the issue!!!!

This happens only when using msbuild 14.

**Comment by valhristov** [comment]
Fixed by https://github.com/SonarSource/sonar-csharp/pull/1445
